### PR TITLE
nixos/direnv: init

### PR DESCRIPTION
**Pull request by infinisil**
###### Description of changes

[nix-direnv](https://github.com/nix-community/nix-direnv) is surprisingly annoying to configure correctly, due to it needing both system-wide adjustments (`nix.conf` changes), but also user-wide adjustments (telling direnv to load nix-direnv using ~/.config/direnv/direnvrc).

This module puts an end to that, allowing a [direnv](https://direnv.net/) + nix-direnv installation with just

```nix
{
  programs.direnv.enable = true;
}
```

Currently this module comes with nix-direnv enabled by default and there's no way to turn it off. If there ever is a need to only enable direnv without nix-direnv we can do that in another PR. I don't need that myself.

Ping @zimbatm @Mic92 @bbenne10 

###### Things done

- [x] Tested the module

### CODE REVIEW

File: nixos/modules/programs/direnv.nix
Comment (ck3d) [inline nixos/modules/programs/direnv.nix:40]: I prefer to use lib.mkDefault instead.

File: pkgs/tools/misc/nix-direnv/default.nix
Comment (zimbatm) [inline pkgs/tools/misc/nix-direnv/default.nix:30]: In that case, the module will also have to handle user settings since DIRENV_CONFIG is also used to locate the `direnv.toml`.

Alternatively, it could make sense to change direnv so it also loads all the libraries in 
`share/direnv/lib/*.sh`, following the XDG spec?

Comment (bbenne10) [reply pkgs/tools/misc/nix-direnv/default.nix:30]: I think that, given that direnv already reads `${XDG_CONFIG_HOME:-$HOME/.config}/direnv/lib/*.sh` it makes perfect sense for it to read the equivalent system directory, does it not?

**Comment by bbenne10** [comment]
This generally looks okay.
I would like to see the addition of something like `programs.direnv.initExtra` that allows the user to specify the additional direnv snippets they're using. The way we're overriding `DIRENV_CONFIG` now makes that impossible (and so the additional snippet that I use that changes where caches are stored is not available at all with this configuration method).

I will say though that I'm not completely sure I see the need to introduce this.
Direnv is somewhat explicitly a user-level tool.
nix-direnv changes this user-level tool a bit to enable caching.
That caching is still useful even if they're occasionally garbage collected without the user knowing.

All of this to say that the system changes are *good* but not wholly necessary to eek out use from nix-direnv.
That being said: I support this change, but I don't think such a simple approach as we see here is as useful in the general sense as @infinisil maybe hopes it'll be?

File: nixos/modules/programs/direnv.nix
Comment (bbenne10) [inline nixos/modules/programs/direnv.nix:25]: As mentioned: This disallows user customization entirely, I think. 
Either we need a change to direnv to address this or we need a change here to give us the ability for the user to specify additional contents.

**Review by bbenne10 — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/nix-direnv/default.nix
Comment (Mic92) [inline pkgs/tools/misc/nix-direnv/default.nix:31]: Could this not be a symlink?

File: pkgs/tools/misc/nix-direnv/default.nix
Comment (infinisil) [reply pkgs/tools/misc/nix-direnv/default.nix:30]: Yeah I think so too, I quote the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html):

> `$XDG_CONFIG_DIRS` defines the preference-ordered set of base directories to search for configuration files in addition to the `$XDG_CONFIG_HOME` base directory. The directories in `$XDG_CONFIG_DIRS` should be seperated with a colon ':'.
>
> If `$XDG_CONFIG_DIRS` is either not set or empty, a value equal to /etc/xdg should be used.

**Comment by infinisil** [comment]
> This generally looks okay. I would like to see the addition of something like `programs.direnv.initExtra` that allows the user to specify the additional direnv snippets they're using. The way we're overriding `DIRENV_CONFIG` now makes that impossible (and so the additional snippet that I use that changes where caches are stored is not available at all with this configuration method).
 
Sounds like this will be addressed by https://github.com/NixOS/nixpkgs/pull/192667#discussion_r979811832

> I will say though that I'm not completely sure I see the need to introduce this. Direnv is somewhat explicitly a user-level tool. nix-direnv changes this user-level tool a bit to enable caching. That caching is still useful even if they're occasionally garbage collected without the user knowing.
> 
> All of this to say that the system changes are _good_ but not wholly necessary to eek out use from nix-direnv. That being said: I support this change, but I don't think such a simple approach as we see here is as useful in the general sense as @infinisil maybe hopes it'll be?

I had a bit of trouble understanding your concern, but I believe you're trying to say that the `nix.settings` are somewhat optional in this module, since they only affect GC relating to nix-direnv. And if we don't need to set `nix.settings`, the module can be entirely user-level, like a home-manager module (of which there already is one [here](https://github.com/nix-community/home-manager/blob/master/modules/programs/direnv.nix)).

Here's some arguments for why I think such a module is useful:
- It is a main feature for nix-direnv to prevent shells from being garbage-collected, see the first couple sentences of [its readme](https://github.com/nix-community/nix-direnv#nix-direnv). Users of nix-direnv expect that feature to work
- Currently if one wants to use direnv + nix-direnv and prevent GC, there is no easy way to do this in a single step. The [NixOS installation instructions](https://github.com/nix-community/nix-direnv#via-configurationnix-in-nixos) for nix-direnv require a manual step of editing `~/.direnvrc`, while the home-manager install instructions require a separate step for adding the correct `nix.conf` settings.
- Personal anecdote: When I installed my NixOS configuration onto a new machine, I was confused to find that my direnv shells were not cached with nix-direnv anymore. Turns out I manually added `~/.direnvrc` (as described in the NixOS installation guide for nix-direnv), which I of course forgot to copy to my new machine (and also every separate user). Preventing such situations is a core selling-point of NixOS.
- If a program can be configured both on a system-level (meaning for all users) and on a user-level (overriding the system-level config), then it makes sense for there to be a NixOS module, because one might want to provide a default for all users. This is also often seen with user-level systemd services provided by the system.

File: nixos/modules/programs/direnv.nix
Comment (ck3d) [reply nixos/modules/programs/direnv.nix:40]: Thanks for the detailed explanation. I did not know that this setting is highly recommended for direnv.

File: pkgs/tools/misc/nix-direnv/default.nix
Comment (infinisil) [reply pkgs/tools/misc/nix-direnv/default.nix:30]: I created a direnv PR for this: https://github.com/direnv/direnv/pull/990

File: nixos/modules/programs/direnv.nix
Comment (infinisil) [reply nixos/modules/programs/direnv.nix:40]: I guess it would only break the GC-rooting functionality of nix-direnv though, not the caching mechanism overall. Relates to https://github.com/NixOS/nixpkgs/pull/192667#issuecomment-1262652503 as well

**Comment by bbenne10** [comment]
> I had a bit of trouble understanding your concern, but I believe you're trying to say that the `nix.settings` are somewhat optional in this module, since they only affect GC relating to nix-direnv. And if we don't need to set `nix.settings`, the module can be entirely user-level, like a home-manager module (of which there already is one [here](https://github.com/nix-community/home-manager/blob/master/modules/programs/direnv.nix)).

This is exactly it.

There are already about 5 ways to install nix-direnv.
This adds a 6th that must additionally be supported.
We already spend quite a bit of debugging time figuring out what version of nix-direnv someone has and how it was installed (as each installation procedure has different pitfalls), and this will (maybe) make this just a bit harder.

**Comment by infinisil** [comment]
> There are already about 5 ways to install nix-direnv. This adds a 6th that must additionally be supported. We already spend quite a bit of debugging time figuring out what version of nix-direnv someone has and how it was installed (as each installation procedure has different pitfalls), and this will (maybe) make this just a bit harder.

No this PR doesn't add another, it will allow you to replace [the existing instructions for NixOS](https://github.com/nix-community/nix-direnv#via-configurationnix-in-nixos) with much simpler ones, namely

```nix
{
  programs.direnv.enable = true;
}
```

**Comment by bbenne10** [comment]
> No this PR doesn't add another, it will allow you to replace [the existing instructions for NixOS](https://github.com/nix-community/nix-direnv#via-configurationnix-in-nixos) with much simpler ones, namely
> 
> ```nix
> {
>   programs.direnv.enable = true;
> }
> ```

I will have to describe the old behavior and the new behavior until acceptance of the new version is high enough. Likely well beyond the 22.11 release, as old installs may not be updated immediately.

Do know that I have said I support this change once appropriate changes are made to direnv as to allow the user to provide augmentations to the direnv stdlib. I am holding off approval until direnv/direnv#990 is approved, merged, released and updated in nixpkgs.

**Comment by infinisil** [comment]
@bbenne10 Awesome that sounds great. You can also just keep everything as-is for now, the old way of installing it continues working the same way.

File: nixos/modules/programs/direnv.nix
Comment (SuperSandro2000) [inline nixos/modules/programs/direnv.nix:10]: This should mention the nix setting changes.

**Comment by drupol** [comment]
What's the status of this useful PR ?

**Comment by bbenne10** [comment]
direnv/direnv#990 is still open. I still haven't signed off on it. This PR would still artificially limit the ability of users to override direnv stdenv functions when using this installation mechanism.

that is to say: no - there's no change, as the referenced PR is stagnate right now too. If we can get that merged, I'm happy to assign here once the appropriate changes are made to allow `extraConfig` or such to propagate into the nix installed direnv / nix-direnv environments.

**Comment by jian-lin** [comment]
FYI, https://github.com/NixOS/nixpkgs/pull/241528

**Pull request closed without merging**
